components/timezone/src/provider/names.rs

```diff
@@ -0,0 +1,207 @@
+///     (NormalizedTimeZoneIdStr::from_str("Asia/Kolkata"), 22),
+///     (NormalizedTimeZoneIdStr::from_str("Europe/Berlin"), 33),
+/// ]
+/// .into_iter()
+/// .collect();
+///
+/// let key_approx = NormalizedTimeZoneIdStr::from_str("europe/berlin");
+/// let key_exact = NormalizedTimeZoneIdStr::from_str("Europe/Berlin");
+///
+/// // Strict lookup:
+/// assert_eq!(None, map.get_copied(key_approx));
+/// assert_eq!(Some(33), map.get_copied(key_exact));
+///
+/// // Loose lookup:
+/// assert_eq!(Some(33), map.get_copied_by(|u| u.cmp_loose(key_approx)));
+/// assert_eq!(Some(33), map.get_copied_by(|u| u.cmp_loose(key_exact)));
+/// ```
+#[derive(PartialEq, Eq)] // VarULE wants these to be byte equality
+#[derive(Debug, VarULE)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize))]
+#[repr(transparent)]
+pub struct NormalizedTimeZoneIdStr(UnvalidatedStr);
+
+/// This impl requires enabling the optional `serde` Cargo feature of the `icu_properties` crate
+#[cfg(feature = "serde")]
+impl<'de> serde::Deserialize<'de> for Box<NormalizedTimeZoneIdStr> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        <Box<UnvalidatedStr>>::deserialize(deserializer).map(NormalizedTimeZoneIdStr::cast_box)
+    }
+}
+
+/// This impl requires enabling the optional `serde` Cargo feature of the `icu_properties` crate
+#[cfg(feature = "serde")]
+impl<'de, 'a> serde::Deserialize<'de> for &'a NormalizedTimeZoneIdStr
+where
+    'de: 'a,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        <&UnvalidatedStr>::deserialize(deserializer).map(NormalizedTimeZoneIdStr::cast_ref)
+    }
+}
+
+impl<'a> ZeroMapKV<'a> for NormalizedTimeZoneIdStr {
+    type Container = VarZeroVec<'a, NormalizedTimeZoneIdStr>;
+    type Slice = VarZeroSlice<NormalizedTimeZoneIdStr>;
+    type GetType = NormalizedTimeZoneIdStr;
+    type OwnedType = Box<NormalizedTimeZoneIdStr>;
+}
+
+/// The Ord/PartialOrd impl will sort things using strict equality, but in such a way that all loose-equal items
+/// will sort into the same area, such that a map can be searched for both strict and loose equality.
+impl PartialOrd for NormalizedTimeZoneIdStr {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+/// The Ord impl will sort things using strict equality, but in such a way that all loose-equal items
+/// will sort into the same area, such that a map can be searched for both strict and loose equality.
+impl Ord for NormalizedTimeZoneIdStr {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let cmp = self.cmp_loose(other);
+        // When loose equality holds, fall back to strict equality
+        if cmp == Ordering::Equal {
+            self.0.cmp(&other.0)
+        } else {
+            cmp
+        }
+    }
+}
+
+impl NormalizedTimeZoneIdStr {
```

Comment:
You have `Box` and `&` constructors, but don't we need `Cow` to keep postcard alloc-free while allowing JSON?

I guess because this is only used in a ZV, postcard uses ULE instead of Serialize, but this should still work correctly in case someone uses this in a non-ULE location.

Reply:
More constructors can be added on an as-needed basis; I wanted to add the ones that are reachable. In fact I should probably double check if there are any I can delete.

Reply:
/s/constructors/serde impls